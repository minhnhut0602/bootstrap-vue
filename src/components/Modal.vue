<template>
    <div
        :id="id"
        ref="modal"
        class="modal"
    >
        <div class="modal-dialog">
            <div class="modal-content">
                <div
                    v-if="!!$slots.header || !!$slots.title || !!title"
                    class="modal-header"
                >
                    <slot name="header">
                        <h5 class="modal-title">
                            <slot name="title">
                                {{ title }}
                            </slot>
                        </h5>

                        <button
                            class="btn-close"
                            data-bs-dismiss="modal"
                            type="button"
                        />
                    </slot>
                </div>

                <div
                    v-if="$slots.default || text"
                    :class="{'modal-body': !noBody}"
                >
                    <slot>
                        {{ text }}
                    </slot>
                </div>

                <div
                    v-if="$slots.footer"
                    class="modal-footer"
                >
                    <slot name="footer" />
                </div>
            </div>
        </div>
    </div>
</template>

<script lang="ts" setup>
import * as bootstrap from 'bootstrap';
import {defineProps, onBeforeUnmount, onMounted, ref} from 'vue';
import {idProps} from '../composables/useId';

defineProps({
    ...idProps,
    noBody: {
        type: Boolean,
        default: false,
    },
    text: {
        type: String,
        default: undefined,
    },
    title: {
        type: String,
        default: undefined,
    },
});

const modal = ref<HTMLElement>();

const bsModal = ref<bootstrap.Modal>();

onMounted(() => {
    bsModal.value = bootstrap.Modal.getInstance(modal.value);
});

onBeforeUnmount(() => {
    if (bsModal.value) {
        bsModal.value.dispose();
    }
});
</script>
