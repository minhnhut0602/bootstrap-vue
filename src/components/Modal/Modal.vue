<template>
    <div
        :id="id"
        ref="modal"
        class="modal"
    >
        <div class="modal-dialog">
            <div class="modal-content">
                <div
                    v-if="!!$slots.header || !!$slots.title || !!title"
                    class="modal-header"
                >
                    <slot name="header">
                        <h5 class="modal-title">
                            <slot name="title">
                                {{ title }}
                            </slot>
                        </h5>

                        <button
                            class="btn-close"
                            data-bs-dismiss="modal"
                            type="button"
                        />
                    </slot>
                </div>

                <div
                    v-if="$slots.default || body"
                    :class="{'modal-body': !noBody}"
                >
                    <slot>
                        {{ body }}
                    </slot>
                </div>

                <div
                    v-if="$slots.footer"
                    class="modal-footer"
                >
                    <slot name="footer" />
                </div>
            </div>
        </div>
    </div>
</template>

<script lang="ts" setup>
import * as bootstrap from 'bootstrap';
import {onBeforeUnmount, onMounted, ref} from 'vue';
import {idProps} from '../../composables/useId';

defineProps({
    body: {
        type: String,
        default: undefined,
    },
    ...idProps,
    noBody: {
        type: Boolean,
        default: false,
    },
    title: {
        type: String,
        default: undefined,
    },
});

const modal = ref<HTMLElement>();

const bsModal = ref<bootstrap.Modal>();

onMounted(() => {
    bsModal.value = bootstrap.Modal.getInstance(modal.value);
});

onBeforeUnmount(() => {
    if (bsModal.value) {
        bsModal.value.dispose();
    }
});
</script>
