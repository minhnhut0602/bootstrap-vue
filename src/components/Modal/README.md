# Modal

```vue
<Modal hideCloseBtn>
    <form>
        ...
    </form>

    <template #footer>
        <Button>Submit</Button>
    </template>
</Modal>
```

## Imported Props

| Prop name | Description                   | Type                                       | Values | Default |
| --------- | ----------------------------- | ------------------------------------------ | ------ | ------- |
| id        | The id used for the component | string                                     | -      | `uuid`  |
| size      | The size of the component     | [Size](../../composables/useSize) (string) | -      | 'md'    |

## Props

| Prop name     | Description                                         | Type                  | Values | Default   |
| ------------- | --------------------------------------------------- | --------------------- | ------ | --------- |
| body          | The body text of the modal                          | string                | -      | undefined |
| headerClass   | The classes that will be forwarded to the header    | string\|array\|object | -      | undefined |
| hideCloseBtn  | Defines if the close button should be hidden        | boolean               | -      | false     |
| modalCentered | Defines if the modal should be vertically centered  | boolean               | -      | false     |
| noBody        | Defines if the `modal-body` class should be omitted | boolean               | -      | false     |
| title         | The title text of the modal                         | string                | -      | undefined |

## Slots

| Name    | Description                            | Bindings |
| ------- | -------------------------------------- | -------- |
| header  | Displays the header of the modal       |          |
| title   | Displays the header title of the modal |          |
| default | Displays the body of the modal         |          |
| footer  | Displays the footer of the modal       |          |
