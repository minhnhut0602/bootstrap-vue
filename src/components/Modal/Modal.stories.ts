
import Btn from '@/components/Button';
import Modal from '.';
import {modalSizes} from '@/composables/useModalSize';

export default {
    title: 'Components/Modal',
    component: Modal,
    argTypes: {
        modalCentered: {
            control: {type: 'boolean'},
        },
        modalSize: {
            control: {type: 'select'},
            options: modalSizes,
        },
    },
};

const template = (args) => ({
    components: {Btn, Modal},
    setup: () => ({args}),
    template: `
        <Modal
            class="d-block"
            title="Modal title"
            text="Modal body text goes here."
            v-bind="args"
        >
            <template #footer>
                <Btn variant="secondary">
                    Close
                </Btn>

                <Btn>
                    Save changes
                </Btn>
            </template>
        </Modal>
    `,
});

export const Default = template.bind({});

export const Demo = (args) => ({
    components: {Btn, Modal},
    setup: () => ({args}),
    template: `
        <Btn
            data-bs-toggle="modal"
            data-bs-target="#modal"
        >
            Launch demo modal
        </Btn>

        <Modal
            id="modal"
            title="Modal title"
            text="Modal body text goes here."
            v-bind="args"
        >
            <template #footer>
                <Btn
                    data-bs-dismiss="modal"
                    variant="secondary"
                >
                    Close
                </Btn>

                <Btn>
                    Save changes
                </Btn>
            </template>
        </Modal>
    `,
});
