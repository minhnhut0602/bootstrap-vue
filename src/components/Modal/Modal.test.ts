import * as bootstrap from 'bootstrap';
import Modal from '.';
import {shallowMount} from '@vue/test-utils';

jest.mock('bootstrap', () => ({
    Modal: {
        getInstance: jest.fn(),
    },
}));

const textPropTest = (propName: string) => {
    it(`renders ${propName}`, () => {
        const text = `I am a ${propName}`;

        const wrapper = shallowMount(Modal, {
            props: {[propName]: text},
        });

        expect(wrapper.find(`.modal-${propName}`).text()).toBe(text);
    });
};

describe('template', () => {
    componentRenderTest(Modal, {
        // Force an id to prevent snapshots from failing because of uuid.
        props: {id: 'dont-change-please'},
    });

    componentSlotRenderTest(Modal, 'header');

    componentSlotRenderTest(Modal, 'title');

    componentSlotRenderTest(Modal);

    componentSlotRenderTest(Modal, 'footer');

    ['title', 'body'].forEach(textPropTest);
});

describe('onMounted', () => {
    it('sets bsModal', async() => {
        const wrapper = await shallowMount(Modal);

        expect(bootstrap.Modal.getInstance).toBeCalledWith(wrapper.vm.$refs.modal);

        // TODO: Unable to assert bsModal value.
    });
});

// TODO: Unable to assert dispose call on bsModal.
