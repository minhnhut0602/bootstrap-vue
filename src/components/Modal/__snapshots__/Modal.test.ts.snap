// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template renders default 1`] = `
"<div id=\\"dont-change-please\\" class=\\"modal\\">
  <div class=\\"modal-dialog\\">
    <div class=\\"modal-content\\">
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
  </div>
</div>"
`;
